### DISCUSSION TITLE
Question on Memory-driven Approximation Simulation and User-defined Memory Threshold

### DISCUSSION THREAD

**Discussion by rentristandelacruz** [munich-quantum-toolkit/ddsim#246] [Q&A] [answered]
I am reading the paper "_As accurate as needed, as efficient as possible: Approximations in DD-based quantum circuit simulation_" under the "_Memory-driven Approximation_" section the paper mentions the following: 

"_...size of the decision diagram representing the state is compared to the user-defined threshold. If that threshold is exceeded, the quantum state is approximated targeting the single-round fidelity (which is also given by the user)_"

Running the "./ddsim_simple -h" shows the following approximation-related options:

- `--approx_when 'fidelity' / 'memory'`
- `--approx_state`
- `--step_fidelity arg`
- `--steps arg`

I'm assuming I can use the option `--approx_when 'memory'` to do memory-driven approximation but I can't find the option to specify that user-defined memory threshold mentioned in the paper. Does this option exist? Maybe this threshold option has already be automated by reading the machine/computer's available memory and using some fraction of it as the memory threshold?

Thanks!

**Comment by burgholzer** [comment]
Many thanks for opening this discussion. I believe @hillmich can best answer that as he was the one implementing the feature.

**Comment by hillmich** [comment, marked as answer]
The option `memory` selects the memory-driven approximation, that's correct. The threshold is determined automatically by comparing the number of nodes in the DD to the GarbageCollection limit. If the number is bigger than this limit, approximation is performed with the `--step_fidelity` fidelity.

Using the available memory (or fraction thereof) as limit has the disadvantages that (i) the DD can still grow larger than this limit during the application of an operation and (ii) if the DD occupies several GiB of memory performance will already suffer.
